examples/schelling_segregation/src/main.cu

```diff
@@ -0,0 +1,357 @@
+    const unsigned int my_y = FLAMEGPU->getVariable<unsigned int, 2>("pos", 1);
+
+    unsigned int same_type_neighbours = 0;
+    unsigned int diff_type_neighbours = 0;
+
+    // Iterate 3x3 Moore neighbourhood (this does not include the central cell)
+    unsigned int my_type = FLAMEGPU->getVariable<unsigned int>("type");
+    for (auto &message : FLAMEGPU->message_in.wrap(my_x, my_y)) {
+        int message_type = message.getVariable<unsigned int>("type");
+        same_type_neighbours += my_type == message_type;
+        diff_type_neighbours += (my_type != message_type) && (message_type != UNOCCUPIED);
+    }
+
+    int isHappy = (static_cast<float>(same_type_neighbours) / (same_type_neighbours + diff_type_neighbours)) > THRESHOLD;
+    FLAMEGPU->setVariable<unsigned int>("happy", isHappy);
+    unsigned int my_next_type = ((my_type != UNOCCUPIED) && isHappy) ? my_type : UNOCCUPIED;
+    FLAMEGPU->setVariable<unsigned int>("next_type", my_next_type);
+    FLAMEGPU->setVariable<unsigned int>("movement_resolved", (my_type == UNOCCUPIED) || isHappy);
+    unsigned int my_availability = (my_type == UNOCCUPIED) || (isHappy == 0);
+    FLAMEGPU->setVariable<unsigned int>("available", my_availability);
+
+    return flamegpu::ALIVE;
+}
+
+FLAMEGPU_AGENT_FUNCTION_CONDITION(is_available) {
+    return FLAMEGPU->getVariable<unsigned int>("available");
+}
+FLAMEGPU_AGENT_FUNCTION(output_available_locations, flamegpu::MessageNone, flamegpu::MessageArray) {
+    FLAMEGPU->message_out.setIndex(FLAMEGPU->getThreadIndex());
+    FLAMEGPU->message_out.setVariable<flamegpu::id_t>("id", FLAMEGPU->getID());
+    return flamegpu::ALIVE;
+}
+
+FLAMEGPU_HOST_FUNCTION(count_available_spaces) {
+    FLAMEGPU->environment.setProperty<unsigned int>("spaces_available", FLAMEGPU->agent("agent").count<unsigned int>("available", 1));
+}
+
+FLAMEGPU_AGENT_FUNCTION_CONDITION(is_moving) {
+    bool movementResolved = FLAMEGPU->getVariable<unsigned int>("movement_resolved");
+    return !movementResolved;
+}
+FLAMEGPU_AGENT_FUNCTION(bid_for_location, flamegpu::MessageArray, flamegpu::MessageBucket) {
+    // Select a location
+    unsigned int selected_index = FLAMEGPU->random.uniform<unsigned int>(0, FLAMEGPU->environment.getProperty<unsigned int>("spaces_available") - 1);
+
+    // Get the location at that index
+    const auto& message = FLAMEGPU->message_in.at(selected_index);
+    const flamegpu::id_t selected_location = message.getVariable<flamegpu::id_t>("id");
+
+    // Bid for that location
+    FLAMEGPU->message_out.setKey(selected_location - 1);
+    FLAMEGPU->message_out.setVariable<flamegpu::id_t>("id", FLAMEGPU->getID());
+    FLAMEGPU->message_out.setVariable<unsigned int>("type", FLAMEGPU->getVariable<unsigned int>("type"));
+    return flamegpu::ALIVE;
+}
+
+FLAMEGPU_AGENT_FUNCTION(select_winners, flamegpu::MessageBucket, flamegpu::MessageArray) {
+    // First agent in the bucket wins
```

Comment:
More expensive, but technically more 'fair' to have agents roll a random float which they attach to the message, and the highest/lowest float wins.